File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int hold = -1 * prices[0];
        int cash = 0;
        
        for(int i = 1; i < prices.length; i++){
            cash = Math.max(cash, hold + prices[i] - fee);
            hold = Math.max(hold, cash - prices[i]);
        }
        
        return cash;
    }
}

File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution(object):
    def maxProfit(self, prices, fee):
        buy = -1 * prices[0]
        sell = 0
        
        for price in prices:
            sell = max(sell, buy + price - fee)
            buy = max(buy, sell - price)            
            
        
        return sell
        